quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -231,110 +304,256 @@ instance Semigroup (Actions state) where
- where
-  loop s [] = s
-  loop s ((var := act) : as) = loop (nextState s act var) as
+prune = loop initialAnnotatedState
+  where
+    loop _s [] = []
+    loop s ((var := act) : as)
+      | computePrecondition s act =
+          (var := act) : loop (computeNextState s act var) as
+      | otherwise =
+          loop s as
+
+withStates :: StateModel state => [Step state] -> [(Step state, Annotated state)]
+withStates = loop initialAnnotatedState
+  where
+    loop _s [] = []
+    loop s ((var := act) : as) =
+      (var := act, s) : loop (computeNextState s act var) as
 
-runActions ::
-  forall state m.
-  (StateModel state, RunModel state m) =>
-  Actions state ->
-  PropertyM m (state, Env m)
-runActions = runActionsInState @_ @m initialState
+stateAfter :: StateModel state => Actions state -> Annotated state
+stateAfter (Actions actions) = loop initialAnnotatedState actions
+  where
+    loop s [] = s
+    loop s ((var := act) : as) = loop (computeNextState s act var) as
 
-runActionsInState ::
+runActions ::
   forall state m.
   (StateModel state, RunModel state m) =>
-  state ->
   Actions state ->
-  PropertyM m (state, Env m)
-runActionsInState st (Actions_ rejected (Smart _ actions)) = loop st [] actions
- where
-  loop _s env [] = do
-    unless (null rejected) $
-      monitor (tabulate "Actions rejected by precondition" rejected)
-    return (_s, reverse env)
-  loop s env ((Var n := act) : as) = do
-    pre $ precondition s act
-    ret <- run (perform s act (lookUpVar env))
-    let name = actionName act
-    monitor (tabulate "Actions" [name])
-    let var = Var n
-        s' = nextState s act var
-        env' = (var :== ret) : env
-    monitor (monitoring @state @m (s, s') act (lookUpVar env') ret)
-    b <- run $ postcondition (s, s') act (lookUpVar env) ret
-    assert b
-    loop s' env' as
+  PropertyM m (Annotated state, Env m)
+runActions (Actions_ rejected (Smart _ actions)) = loop initialAnnotatedState [] actions
+  where
+    loop :: Annotated state -> Env m -> [Step state] -> PropertyM m (Annotated state, Env m)
+    loop _s env [] = do
+      unless (null rejected) $
+        monitor (tabulate "Actions rejected by precondition" rejected)
+      return (_s, reverse env)
+    loop s env ((Var n := act) : as) = do
+      pre $ computePrecondition s act
+      ret <- run (perform (underlyingState s) act (lookUpVar env))
+      let name = actionName act
+      monitor (tabulate "Actions" [name])
+      let var = Var n
+          s' = computeNextState s act var
+          env' = (var :== ret) : env
+      monitor (monitoring @state @m (underlyingState s, underlyingState s') act (lookUpVar env') ret)
+      b <- run $ postcondition @state @m (underlyingState s, underlyingState s') act (lookUpVar env) ret
+      assert b
+      loop s' env' as
+
+-- Trics to get HasVariables instances via either TemplateHaskell or Generic
+
+makeHasVarsInstance :: TH.Type -> [Con] -> Q InstanceDec
```

Comment:
Suggestion: having a separate module for `TH` stuff seems to be common practice in order to let people use what they want and not require to bring TH in?

Reply:
Do you need to bring it in if you just depend on a module that defines TH stuff? In general I agree that restructuring this a little bit could help.

Reply:
If you don't use the TH stuff then I think you don't need explicit extension or import so that's not a big deal. But in general, separating things a bit more could help as this module tends to become a bit unwieldy. Perhaps having a separate namespace `Test.QuickCheck.StateModel.TH` ?

Reply:
Yes, I do agree. This was mostly a point of curiosity. I'm planning to move all the variable handling stuff to one module and the TH stuff to another module.